File: java/lesson 3/SeparateChainingHashTable.java
// https://leetcode.com/problems/design-hashmap/

public class SeparateChainingHashTable<Key, Value> implements HashTable<Key, Value> {
    private static final int INITIAL_CAPACITY = 4;
    private int count;
    private int capacity;
    private Node<Key, Value>[] buckets;

    public SeparateChainingHashTable() {
        this(INITIAL_CAPACITY);
    }

    public SeparateChainingHashTable(int capacity) {
        this.capacity = capacity;
        this.buckets = (Node<Key, Value>[]) new Node[capacity];
    }

    public void put(Key key, Value value) {
        if (count >= 10 * capacity) resize(2 * capacity);
        int i = hash(key);
        for (Node<Key, Value> node = buckets[i]; node != null; node = node.next) {
            if (key.equals(node.key)) {
                node.value = value;
                return;
            }
        }
        buckets[i] = new Node<Key, Value>(key, value, buckets[i]);
        count++;
    }

    public Value get(Key key) {
        int i = hash(key);
        for (Node<Key, Value> node = buckets[i]; node != null; node = node.next) {
            if (key.equals(node.key)) {
                return node.value;
            }
        }
        return null;
    }

    public void remove(Key key) {
        int i = hash(key);
        if (buckets[i] == null) return;

        for (Node prev = null, node = buckets[i]; node != null; prev = node, node = node.next) {
            if (key.equals(node.key)) {
                if (prev == null) {
                    buckets[i] = node.next;
                } else {
                    prev.next = node.next;
                }
                count--;
                break;
            }
        }
        if (capacity > INITIAL_CAPACITY && count <= 2 * capacity) resize(capacity / 2);
    }

    private int hash(Key key) {
        return (key.hashCode() & 0x7fffffff) % capacity;
    }

    private void resize(int size) {
        SeparateChainingHashTable<Key, Value> temp = new SeparateChainingHashTable<Key, Value>(size);
        for (Node<Key, Value> item : buckets) {
            for (Node<Key, Value> node = item; node != null; node = node.next) {
                temp.put(node.key, node.value);
            }
        }
        this.buckets = temp.buckets;
        this.capacity = size;
    }

    private class Node<Key, Value> {
        public Key key;
        public Value value;
        public Node<Key, Value> next;

        public Node(Key key, Value value, Node<Key, Value> next) {
            this.key = key;
            this.value = value;
            this.next = next;
        }
    }
}

File: kotlin/lesson 3/SeparateChainingHashTable.kt
// https://leetcode.com/problems/design-hashmap/

public class SeparateChainingHashTable<Key, Value>(var capacity: Int = INITIAL_CAPACITY) : HashTable<Key, Value> {
            
    private var count = 0
    private var buckets = Array<Node<Key,Value>?>(capacity, { null })

    override fun put(key: Key, value: Value) {
        if (count >= 10 * capacity) resize(2 * capacity)
        var i = hash(key)
        var node = buckets[i]
        while (node != null) {
            if (key == node.key) {
                node.value = value
                return
            }
            node = node.next
        }        
        buckets[i] = Node<Key, Value>(key, value, buckets[i])
        count++
    }

    override fun get(key: Key): Value? {
        var node = buckets[hash(key)]
        while (node != null) {
            if (key == node.key) return node.value
            node = node.next
        }        
        return null
    }

    override fun remove(key: Key) {
        var i = hash(key)        
        if (buckets[i] == null) return
        var prev: Node<Key, Value>? = null
        var node = buckets[i]        
        while (node != null) {
            if (key == node.key) {
                if (prev == null) {
                    buckets[i] = node.next
                } else {
                    prev.next = node.next
                }
                count--
                break
            }            
            prev = node
            node = node.next
        }
        
        if (capacity > INITIAL_CAPACITY && count <= 2 * capacity) resize(capacity / 2)
    }

    private fun hash(key: Key): Int {
        return key.hashCode() % capacity
    }

    private fun resize(size: Int) {
        val temp = SeparateChainingHashTable<Key, Value>(size)
        for (bucket in buckets) {
            var node = bucket
            while (node != null) {
                temp.put(node.key, node.value)
                node = node.next
            }
        }
        this.buckets = temp.buckets
        this.capacity = size
    }

    companion object {
        const val INITIAL_CAPACITY = 4
    }
    
    class Node<Key, Value>(var key: Key, var value: Value, var next: Node<Key, Value>? = null)
}
